File: Payload Computer/get_metadata.py
"""
This script will run on startup on the payload computer. When a flag is set through ROS, it will begin 
saving metadata from the USB connection. Note that it requires MAVROS to be running in parallel to work.
Note also that data from each ROS topic is saved in a separate file.

Author: Krishna Makhija
date: Sep 8th 2020
"""

import rospy, time, re
import numpy as np
from os.path import expanduser
from termcolor import colored
from std_msgs.msg import Float32
from threading import Event, Thread
from geometry_msgs.msg import PoseStamped
from termcolor import colored
from mavros_msgs.msg import WaypointReached, WaypointList, PositionTarget
from sensor_msgs.msg import Imu, NavSatFix
from threading import Event
from std_msgs.msg import String
from mavros_msgs.msg import *
from mavros_msgs.srv import *


path            = expanduser("~") + "/"             # data files save path
logs_path       = path + '/catkin_ws/src/Drone-Project-code/logs/metadata/'             
local_pose      = logs_path + time.strftime("%d-%m-%Y_%H-%M-%S_local_pose.log")
global_pos      = logs_path + time.strftime("%d-%m-%Y_%H-%M-%S_global_pose.log")
set_target      = logs_path + time.strftime("%d-%m-%Y_%H-%M-%S_set_target.log")
sdr_drone_temp  = logs_path + time.strftime("%d-%m-%Y_%H-%M-%S_sdr_drone_temp.log")
refresh_rate    = 10.0


def get_timestamp():
    """
    Returns current time for data logging.
    """
    time_now = time.time()
    mlsec = repr(time_now).split('.')[1][:3]
    time_now = time.strftime("%H:%M:%S.{}-%d/%m/%Y".format(mlsec))
    return time_now


def callback_local(data):
    """
    Callback object for drone's local position and timestamp.
    """
    try:
        current_time = get_timestamp()
        local_pose_f_a.write("%s\t%s\t%s\t%s\n" % (current_time, data.pose.position.x,
                                                   data.pose.position.y, data.pose.position.z))
        rospy.sleep(1/refresh_rate)
    except ValueError:
        pass


def callback_setpoint(data):
    """
    Callback object for drone's setpoint position and timestamp.
    """
    try:
        current_time = get_timestamp()
        set_target_f_a.write("%s\t%s\t%s\t%s\n" %
                             (current_time, data.position.x, data.position.y, data.position.z))
        rospy.sleep(1/refresh_rate)
    except ValueError:
        pass


def callback_global(data):
    """
    Callback object for drone's GPS location.
    """
    try:
        current_time = get_timestamp()
        global_pos_f_a.write("%s\t%s\t%s\t%s\n" %(current_time, data.latitude, data.longitude, data.altitude))
        rospy.sleep(1/refresh_rate)
    except ValueError:
        pass


def callback_SDR(data):
    """
    Callback object for SDR temperature.
    """
    try:
        current_time = get_timestamp()
        sdr_drone_temp_f_a.write("%s\t%s\n" %(current_time, data.data))
        rospy.sleep(1/refresh_rate)
    except ValueError, NameError:
        pass


def main_sitl():
    """
    Initiate metadata file and write to it. Apparently, this only works with the SITL sims. Go figure.
    """
    rospy.init_node('get_metadata', anonymous=True)
    rospy.set_param('trigger/metadata', False)
    file = open(metadata, "w+")
    file.write("Timestamp\tLocal Position (x)\tLocal Position (y)\tLocal Position (z)\tSet_target (x)\tSetpoint (y)\tSetpoint (z)\tTemperature\n")
    file.close()
    wp_num = 0
    print(colored('ROS metadata node intialized. Waiting for flag from SDR code to begin saving metadata.', 'green'))
    while not rospy.is_shutdown():
        time.sleep(0.001)
        if rospy.get_param('trigger/metadata') == True:
            print(colored('Saving Waypoint #' + str(wp_num) +
                          ' metadata in ' + str(metadata), 'grey', 'on_white'))
            file = open(metadata, "a+")
            file.write("Waypoint #%s\n" % (wp_num))
            wp_num += 1
            while True:
                current_time = time.strftime("%H%M%S-%d%m%Y")
                print(current_time)
                local_pose = rospy.wait_for_message(
                    '/mavros/local_position/pose', PoseStamped)
                set_target = rospy.wait_for_message(
                    '/mavros/setpoint_raw/target_local', PositionTarget)
                sdr_temp = rospy.wait_for_message('sdr_temperature', Float32)
                file.write("%s\t%s\t%s\t%s\t" % (current_time, local_pose.pose.position.x,
                                                 local_pose.pose.position.y, local_pose.pose.position.z))
                file.write("%s\t%s\t%s\n" % (set_target.position.x,
                                             set_target.position.y, set_target.position.z))

                if rospy.get_param('trigger/metadata') == False:
                    print('Finished saving metadata for this WP.')
                    file.close()
                    break


def main():
    """
    main function for acquiring metadata at waypoints.
    """
    global local_pose_f_a
    global set_target_f_a
    global global_pos_f_a
    global sdr_drone_temp_f_a

    local_pose_f = open(local_pose, "w+")
    local_pose_f.write(
        "Timestamp\tLocal Position (x)\tLocal Position (y)\tLocal Position (z)\n")
    local_pose_f.close()
    set_target_f = open(set_target, "w+")
    set_target_f.write(
        "Timestamp\tSet_target (x)\tSet_target (y)\tSet_target (z)\n")
    set_target_f.close()
    global_pos_f = open(global_pos, "w+")
    global_pos_f.write("Timestamp\tLatitude\tLongitude\tAltitude\n")
    global_pos_f.close()
    sdr_drone_temp_f = open(sdr_drone_temp, "w+")
    sdr_drone_temp_f.write("Timestamp\tSDR Temperature (deg C)")
    sdr_drone_temp_f.close()

    rospy.init_node('get_metadata', anonymous=True)
    rospy.set_param('trigger/metadata', False)
    wp_num = 0
    print(colored('ROS metadata node intialized. Waiting for flag from SDR code to begin saving metadata.', 'green'))
    while not rospy.is_shutdown():
        time.sleep(0.01)
        if rospy.get_param('trigger/metadata') == True:
            current_time = time.strftime("%H%M%S-%d%m%Y")
            print(colored('[' + str(current_time) + ']Saving Waypoint #' + str(wp_num) + ' metadata in ' +
                          str(local_pose) + ' and ' + str(set_target), 'grey', 'on_white'))
            local_pose_f_a = open(local_pose, "a+")
            set_target_f_a = open(set_target, "a+")
            global_pos_f_a = open(global_pos, "a+")
            sdr_drone_temp_f_a = open(sdr_drone_temp, "a+")
            local_pose_f_a.write("Waypoint #%s\n" % (wp_num))
            set_target_f_a.write("Waypoint #%s\n" % (wp_num))
            global_pos_f_a.write("Waypoint #%s\n" % (wp_num))
            sdr_drone_temp_f_a.write("Waypoint #%s\n" % (wp_num))
            wp_num += 1
            rospy.Subscriber('/mavros/local_position/pose',
                                PoseStamped, callback_local)
            rospy.Subscriber('/mavros/setpoint_raw/target_local',
                                PositionTarget, callback_setpoint)
            rospy.Subscriber('/mavros/global_position/global', NavSatFix, callback_global)
            rospy.Subscriber('/sdr_temperature', Float32, callback_SDR)
            while True:
                if rospy.get_param('trigger/metadata') == False:
                    print('Finished saving metadata for this WP.')
                    local_pose_f_a.close()
                    set_target_f_a.close()
                    global_pos_f_a.close()
                    sdr_drone_temp_f_a.close()
                    break


if __name__ == '__main__':
    main()

File: src/get_metadata.cpp
/* 
* C++ version of get_metadata.py
* When a WP is reached a flag is set ("trigger/sequence") which begins calibration from the 
* cal_sequence script. That script also sets the trigger/metadata flag which causes this 
* program to begin saving metadata from the FCU and SDR. 

* author: KM
* date: 28th Aug 2021
*/

#include "ros/ros.h"
#include <iostream>
#include <stdio.h>
#include <fstream> 
#include <unistd.h>
#include <time.h>
#include <sys/time.h>
#include <string.h>
#include <sys/types.h>
#include <pwd.h>
#include <iomanip>          // std::setprecision()
#include "sensor_msgs/Imu.h"
#include "nav_msgs/Odometry.h"
#include "geometry_msgs/PoseStamped.h"
#include "mavros_msgs/PositionTarget.h"
#include "sensor_msgs/NavSatFix.h"
#include "std_msgs/Float32.h"

// Define global variables
double refresh_rate = 20.0;         // sensor refresh rate set by mavros (see mission.sh)
bool seq_flag;                      // sequene flag initialization b/c I do not understand C++
bool met_flag;                      // metadata flag initialization b/c I do not understand C++
struct passwd *pw = getpwuid(getuid());
const char *homedir = pw->pw_dir;   // get path to home directory
std::string local_pose;             
std::string global_pose;            // these will point to the metadata filenames
std::string sdr_d_temp;
std::ofstream local_pose_f;
std::ofstream global_pose_f;        // declare ofsteam objects for writing to files later
std::ofstream sdr_d_temp_f;         //      using the callback sub functions

//? get current time and format for timestamping metadata
std::string get_timestamp(){
    struct timeval tmnow;
    struct tm *tm;
    char timestamp[30], usec_time[12], day[30];         // declare empty strings
    gettimeofday(&tmnow, NULL);                         // get time
    tm = localtime(&tmnow.tv_sec);                      // convert to readable time
    strftime(timestamp, 30, "%H:%M:%S", tm);            // format time upto sec resolution
    strcat(timestamp, ".");                             // add decimal point
    sprintf(usec_time, "%d", (int)tmnow.tv_usec);       // get time in ms
    strcat(timestamp, usec_time);                       // add time in ms res to time string
    strftime(day, 30, "-%d/%m/%Y", tm);                 // repeat the above to
    strcat(timestamp, day);                             //   have day info added as well
    return timestamp;
}

//? create a different timestamp for metadata filenames
std::string get_filename(){
    struct timeval tmnow;
    struct tm *tm;
    char namestamp[45];                                 // declare empty strings
    gettimeofday(&tmnow, NULL);                         // get time
    tm = localtime(&tmnow.tv_sec);                      // convert to readable time
    strftime(namestamp, 45, "%d-%m-%Y_%H-%M-%S", tm);   // translate to convenient format
    return namestamp;
}

//? close files after writing each line. is there a better way to do it?
void close_files(){
    local_pose_f.close();
    global_pose_f.close();
    sdr_d_temp_f.close();
}

//? define callback functions. these will write metadata to files.
//* IMU pose data
void callback_local(const geometry_msgs::PoseStamped::ConstPtr& msg){
    std::string current_time = get_timestamp();
    local_pose_f.open(local_pose, std::ios_base::app);
    local_pose_f << current_time << "\t" << msg->pose.position.x << "\t" 
                    << msg->pose.position.y << "\t" << msg->pose.position.z 
                    << std::setprecision(12) << std::endl;
}

//* GPS raw data
void callback_global(const sensor_msgs::NavSatFix::ConstPtr& msg){
    std::string current_time = get_timestamp();
    global_pose_f.open(global_pose, std::ios_base::app);
    global_pose_f << current_time << "\t" << msg->latitude << "\t" 
                    << msg->longitude << "\t" << msg->altitude  
                    << std::setprecision(12) << std::endl; 
}

//* SDR temperature
void callback_SDR(const std_msgs::Float32::ConstPtr& msg){
    std::string current_time = get_timestamp();
    sdr_d_temp_f.open(sdr_d_temp, std::ios_base::app);
    sdr_d_temp_f << current_time << "\t" << msg->data 
                    << std::setprecision(12) << std::endl;
}

int main(int argc, char **argv){
    //? create metadata files and add column headings
    std::string time_now = get_filename();
    std::string logs_path = "/catkin_ws/src/Drone-Project-code/logs/metadata/";
    logs_path = homedir + logs_path;
    std::string &local_ptr = local_pose;
    local_ptr = logs_path + time_now + "_local_pose.log";
    local_pose_f.open(local_pose);
    local_pose_f << "Timestamp\tLocal Position (x)\tLocal Position (y)\tLocal Position (z)\n";
    std::string &global_ptr = global_pose;
    global_ptr = logs_path + time_now + "_global_pose.log";
    global_pose_f.open(global_pose);
    global_pose_f << "Timestamp\tLongitude\tLatitude\tAltitude\n";
    std::string &sdr_d_ptr = sdr_d_temp;
    sdr_d_ptr = logs_path + time_now + "_sdr_d_temp.log";
    sdr_d_temp_f.open(sdr_d_temp);
    sdr_d_temp_f << "Timestamp\tSDR Temperature (deg C)\n";
    close_files();

    //? initiate ROS node
    ros::init(argc, argv, "get_metadata");
    ros::NodeHandle n;
    n.setParam("trigger/metadata", false);
    int wp_num = 0;                                                 // WP number counter
    ROS_INFO("ROS metadata node intialized. Waiting for flag from SDR code to begin saving metadata.");
    ros::Rate r(refresh_rate);
    while (ros::ok()) {
        usleep(1/refresh_rate * 1e6);                               // refresh rate in us
        if (n.param("trigger/metadata", met_flag) == true){         // check if flag set by wp_trigger
            ROS_INFO("Saving Waypoint #%i metadata in ./logs/metadata/", wp_num);
            local_pose_f.open(local_pose, std::ios_base::app);
            global_pose_f.open(global_pose, std::ios_base::app);    // append data to previous
            sdr_d_temp_f.open(sdr_d_temp, std::ios_base::app);      //      ofstream objects
            local_pose_f << "Waypoint #" << wp_num << std::endl;
            global_pose_f << "Waypoint #" << wp_num << std::endl;   // tag new waypoints
            sdr_d_temp_f << "Waypoint #" << wp_num << std::endl;
            close_files();
            wp_num++;
            ros::Subscriber loc_sub = n.subscribe("/mavros/local_position/pose", 1000, callback_local);
            ros::Subscriber glo_sub = n.subscribe("/mavros/global_position/global", 1000, callback_global);
            ros::Subscriber sdr_sub = n.subscribe("/sdr_temperature", 1000, callback_SDR);
            while (ros::ok()){
                ros::spinOnce();                                    // spin once because subscribe                                 
                r.sleep();                                          //   and spin work differently
                close_files();                                      //    in C++ and Python
                if (n.param("trigger/metadata", met_flag) == false){
                    ROS_INFO("Finished saving metadata for this WP.");
                    close_files();
                    loc_sub.shutdown();
                    glo_sub.shutdown();
                    sdr_sub.shutdown();
                    break;
                }
            }
        }
    }
    return 0;
}

